File: src/main/java/dev.toad/msg/MessageType.java
package dev.toad.msg;

public enum MessageType {
  CON(1),
  NON(2),
  ACK(3),
  RESET(4);

  private MessageType(int val) {}

  public String toString() {
    return switch (this) {
      case CON -> "CON";
      case NON -> "NON";
      case ACK -> "ACK";
      case RESET -> "RESET";
      default -> throw new Error();
    };
  }

  public static MessageType fromString(String s) {
    return switch (s.toLowerCase().trim()) {
      case "con" -> CON;
      case "non" -> NON;
      case "ack" -> ACK;
      case "reset" -> RESET;
      default -> throw new Error();
    };
  }
}

File: glue/src/message_type.rs
use toad_jni::java;
use toad_msg::Type;

pub struct MessageType(java::lang::Object);

java::object_newtype!(MessageType);
impl java::Class for MessageType {
  const PATH: &'static str = package!(dev.toad.msg.MessageType);
}

impl MessageType {
  pub fn new(env: &mut java::Env, ty: Type) -> Self {
    static FROM_STRING: java::StaticMethod<MessageType, fn(String) -> MessageType> =
      java::StaticMethod::new("fromString");

    let str = match ty {
      | Type::Con => "CON",
      | Type::Non => "NON",
      | Type::Ack => "ACK",
      | Type::Reset => "RESET",
    };

    FROM_STRING.invoke(env, str.to_string())
  }

  pub fn to_toad(&self, env: &mut java::Env) -> Type {
    static TO_STRING: java::Method<MessageType, fn() -> String> = java::Method::new("toString");

    match TO_STRING.invoke(env, self).trim().to_uppercase().as_str() {
      | "CON" => Type::Con,
      | "NON" => Type::Non,
      | "ACK" => Type::Ack,
      | "RESET" => Type::Reset,
      | o => panic!("malformed message type {}", o),
    }
  }
}
